Allow iterations (work factor) to be provided to Account#encrypt

stefanmendoza opened the pull request:
## What was wrong?

* Resolves #48
* **(EDIT: Moved to #53)**`pytest-xdist 1.25.0 has requirement pytest>=3.6.0, but you'll have pytest 3.3.2 which is incompatible.` ([1.24.1](https://github.com/pytest-dev/pytest-xdist/blob/v1.24.1/setup.py#L3) vs. [1.25.0](https://github.com/pytest-dev/pytest-xdist/blob/v1.25.0/setup.py#L3))

## How was it fixed?

* Expose `iterations` as an optional param to `Account#encrypt`
* **(EDIT: Moved to #53)** Uplift `pytest` to `>=3.6.0` and fix the resulting breaking changes in tests accordingly
* Tests added for providing the work factor to Account.encrypt

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://images.vice.com/vice/images/articles/meta/2015/05/28/why-do-i-want-to-crush-cute-animals-1432782372.jpg)

stefanmendoza commented:
On master:
```
$ pytest tests/
....
pkg_resources.VersionConflict: (pytest 3.3.2 (/Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.7/site-packages), Requirement.parse('pytest>=3.6.0'))

....

pluggy.PluginValidationError: Plugin 'xdist' could not be loaded: (pytest 3.3.2 (/Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.7/site-packages), Requirement.parse('pytest>=3.6.0'))!
```

stefanmendoza commented on tests/core/test_accounts.py:
```diff
@@ -434,44 +419,83 @@ def test_eth_account_recover_transaction_from_eth_test(acct, transaction):
-@pytest.mark.parametrize(
-    'private_key, password, kdf, expected_decrypted_key, expected_kdf',
-    [
+def encrypt_mark_params():
```
> tests/core/test_accounts.py:441: in <module>
    web3js_key(),
E   _pytest.warning_types.RemovedInPytest4Warning: Fixture "web3js_key" called directly. Fixtures are not meant to be called directly, are created automatically when test functions request them as parameters. See https://docs.pytest.org/en/latest/fixture.html for more information.

See: https://github.com/pytest-dev/pytest/issues/4112#issuecomment-439039575

AndreMiras replied:
Minor suggestion: Yep this change seems to make sense, but I generally try not to mix concerns of PR to make sure they get merged nice and easy.
That of course depends on core-dev policies and I'm not sure what's applied here

stefanmendoza replied:
The build is actually failing due to this.

AndreMiras replied:
Oh I didn't know they were failing on warnings, well done then!

stefanmendoza replied:
```
====================================================================================== ERRORS =======================================================================================
___________________________________________________________________ ERROR collecting tests/core/test_accounts.py ____________________________________________________________________
tests/core/test_accounts.py:441: in <module>
    web3js_key(),
E   _pytest.warning_types.RemovedInPytest4Warning: Fixture "web3js_key" called directly. Fixtures are not meant to be called directly, are created automatically when test functions request them as parameters. See https://docs.pytest.org/en/latest/fixture.html for more information.
================================================================================= warnings summary ==================================================================================
venv/lib/python3.7/site-packages/eth_keys/datatypes.py:105
  /Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.7/site-packages/eth_keys/datatypes.py:105: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    class BaseKey(ByteString, collections.Hashable):

eth_account/account.py:1
  /Users/stefanmendoza/Github/Forks/ethereum/eth-account/eth_account/account.py:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import (

venv/lib/python3.7/site-packages/rlp/sedes/raw.py:6
  /Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.7/site-packages/rlp/sedes/raw.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Sequence

venv/lib/python3.7/site-packages/rlp/lazy.py:1
  /Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.7/site-packages/rlp/lazy.py:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Iterable, Sequence

venv/lib/python3.7/site-packages/attrdict/mixins.py:5
  /Users/stefanmendoza/Github/Forks/ethereum/eth-account/venv/lib/python3.7/site-packages/attrdict/mixins.py:5: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Mapping, MutableMapping, Sequence

-- Docs: https://docs.pytest.org/en/latest/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================== 5 warnings, 1 error in 0.59 seconds ========================================================================
```

stefanmendoza commented:
`master` (with `pytest-xdist==1.24.1`):
```
============================================================================= slowest 10 test durations =============================================================================
2.22s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str_provided_kdf]
2.19s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str]
2.17s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str]
2.17s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str_provided_kdf]
2.16s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[eth_keys.datatypes.PrivateKey]
2.14s call     tests/core/test_accounts.py::test_eth_account_encrypt[eth_keys.datatypes.PrivateKey]
0.05s call     tests/core/test_accounts.py::test_eth_account_recover_transaction_from_eth_test[transaction2]
0.05s call     tests/core/test_accounts.py::test_eth_account_sign_transaction_from_eth_test[transaction0]
0.04s call     tests/core/test_accounts.py::test_eth_account_sign_transaction_from_eth_test[transaction2]
0.03s call     tests/core/test_accounts.py::test_eth_account_sign_transaction[web3js_hex_str_example]
============================================================================ 62 passed in 14.45 seconds =============================================================================
```

This branch (with the new tests that override iterations excluded):
```
============================================================================= slowest 10 test durations =============================================================================
2.08s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str_provided_kdf]
2.05s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str]
2.04s call     tests/core/test_accounts.py::test_eth_account_encrypt[eth_keys.datatypes.PrivateKey]
1.94s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str]
1.93s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[eth_keys.datatypes.PrivateKey]
1.91s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str_provided_kdf]
0.04s call     tests/core/test_accounts.py::test_eth_account_sign_transaction_from_eth_test[transaction1]
0.04s call     tests/core/test_accounts.py::test_eth_account_sign_transaction_from_eth_test[transaction2]
0.03s call     tests/core/test_accounts.py::test_eth_account_sign_transaction_from_eth_test[transaction0]
0.03s call     tests/core/test_accounts.py::test_eth_account_recover_signature_bytes[test_sig_bytes_standard_v]
====================================================================== 62 passed, 5 warnings in 13.27 seconds =======================================================================
```

So it looks like performance is still comparable.

stefanmendoza commented on eth_account/account.py:
```diff
@@ -145,6 +145,7 @@ def encrypt(cls, private_key, password, kdf=None):
         :type private_key: hex str, bytes, int or :class:`eth_keys.datatypes.PrivateKey`
         :param str password: The password which you will need to unlock the account in your client
         :param str kdf: The key derivation function to use when encrypting your private key
+        :param str iterations: The work factor for the key derivation function
```
I opted for `iterations` instead of `work_factor` to align with the param on the public `create_keyfile_json` function this will call.

stefanmendoza commented on eth_account/account.py:
```diff
@@ -191,7 +192,11 @@ def encrypt(cls, private_key, password, kdf=None):
+        if iterations is None:
+            return create_keyfile_json(key_bytes, password_bytes, kdf=kdf)
+        else:
+            return create_keyfile_json(key_bytes, password_bytes, kdf=kdf, iterations=iterations)
```
I decided to pass `iterations` to the method as-is (vs. another conditional block to not pass this if `iterations` is `None`) since [`create_keyfile_json`](https://github.com/ethereum/eth-keyfile/blob/v0.5.1/eth_keyfile/keyfile.py#L39) calls `_create_v3_keyfile_json` which [has a check if this is `None`](https://github.com/ethereum/eth-keyfile/blob/v0.5.1/eth_keyfile/keyfile.py#L87-L88)]. If we want more safety in case that goes away since that's an internal check in a private method, we can conditionally pass it as well similarly to how we do with the KDF.

AndreMiras replied:
What you say make sense here, perhaps it would be even better to have it as a code comment rather than a PR comment. That way would be IMO more future proof.
Also you may want to enforce/unittest this behaviour via `unittest.mock`  (checking actual mock call) to make sure this feature is covered

stefanmendoza replied:
Sure, I'll have to look into how to test that since the method is in `eth_keyfile` and not the `Account` object.

pipermerriam replied:
I *think* I'm :-1: on this precaution.  I understand the reasoning but given that the `create_keyfile_json` API is public we'd issue a major version bump for any breaking API change.

stefanmendoza replied:
So you want me to just pass the iterations to the param regardless of if it's `None`?

pipermerriam replied:
Yes, directly passing `None` in should be equivalent.

stefanmendoza replied:
Fixed! 65864f3

AndreMiras commented on eth_account/account.py:
```diff
@@ -145,6 +145,7 @@ def encrypt(cls, private_key, password, kdf=None):
         :type private_key: hex str, bytes, int or :class:`eth_keys.datatypes.PrivateKey`
         :param str password: The password which you will need to unlock the account in your client
         :param str kdf: The key derivation function to use when encrypting your private key
+        :param str iterations: The work factor for the key derivation function
```
Minor: docstring typo, the type should be `int` not `str`

stefanmendoza replied:
Good catch!

Fixed: 91ad49a

AndreMiras commented on setup.py:
```diff
@@ -7,7 +7,7 @@
 extras_require={
     'test': [
-        "pytest==3.3.2",
+        "pytest>=3.6.0",
```
Is this a strong requirement of this pull request? I generally try to touch the minimum when I PR, but maybe that one is mandatory

stefanmendoza replied:
Yeah. `master` is failing due to a new release:

https://github.com/ethereum/eth-account/pull/52#issuecomment-450754364

`pytest-xdist 1.25.0 has requirement pytest>=3.6.0, but you'll have pytest 3.3.2 which is incompatible.` ([1.24.1](https://github.com/pytest-dev/pytest-xdist/blob/v1.24.1/setup.py#L3) vs. [1.25.0](https://github.com/pytest-dev/pytest-xdist/blob/v1.25.0/setup.py#L3))

AndreMiras replied:
Great, I was suspecting something like this, but didn't check, sorry. Thanks for your reply :beers:

AndreMiras commented on tests/core/test_accounts.py:
```diff
@@ -434,44 +419,83 @@ def test_eth_account_recover_transaction_from_eth_test(acct, transaction):
         ),
-    ],
-    ids=['hex_str', 'hex_str_provided_kdf', 'eth_keys.datatypes.PrivateKey']
+        (
```
Then would it make sense to decrease the number of iterations?

stefanmendoza replied:
I could do that. I just doubled whatever the default was for each KDF. I'll adjust them and see how it affects the duration.

stefanmendoza replied:
Prior to reducing:
```
============================================================================= slowest 10 test durations =============================================================================
4.78s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str_scrypt_provided_iterations]
4.71s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str_pbkdf2_provided_iterations]
4.38s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str_scrypt_provided_iterations]
3.89s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str_pbkdf2_provided_iterations]
2.39s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str]
2.22s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str_provided_kdf]
2.18s call     tests/core/test_accounts.py::test_eth_account_encrypt[eth_keys.datatypes.PrivateKey]
1.98s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str_provided_kdf]
1.91s call     tests/core/test_accounts.py::test_eth_account_encrypt[hex_str]
1.91s call     tests/core/test_accounts.py::test_eth_account_prepared_encrypt[eth_keys.datatypes.PrivateKey]
====================================================================== 95 passed, 5 warnings in 31.64 seconds =======================================================================
```

After reducing:
```
...
```

Commit: daea1fd

AndreMiras commented:
Looking very good to me, thanks!

stefanmendoza commented:
@AndreMiras introduced a PR to fix master / the pytest issue. Will rebase this after. I agree that we can separate concerns here.

https://github.com/ethereum/eth-account/pull/53

cc @pipermerriam  / @carver  / @davesque

stefanmendoza commented:
Thanks @pipermerriam!

stefanmendoza commented:
@pipermerriam any thoughts on this?
https://github.com/ethereum/eth-account/pull/52#discussion_r245496497

carver commented on tests/core/test_accounts.py:
```diff
@@ -428,48 +428,74 @@ def encrypt_mark_params():
+            public_key,
+            password,
+            'scrypt',
+            1024,
```
At least two things would improve these tests:

1. run an example where `kdf` is `None` and `iterations` is set.
2. verify the iterations value in the keyfile (note that the iterations value output is sensitive to which kdf is used)

stefanmendoza replied:
Rearranged the tests a bit. First and second tests are just the basic private key ones, the rest are messing with the KDF and iterations. The fourth test (new one I added) is the one you're asking about:
* https://github.com/ethereum/eth-account/pull/52/commits/aa9182811a67a673dc428f1f37ea41fad14754fc#diff-1de692d823115945c2232e1aa3066811R463
* https://github.com/ethereum/eth-account/pull/52/commits/aa9182811a67a673dc428f1f37ea41fad14754fc#diff-1de692d823115945c2232e1aa3066811R540

And added the iterations validation [here](https://github.com/ethereum/eth-account/pull/52/commits/aa9182811a67a673dc428f1f37ea41fad14754fc#diff-1de692d823115945c2232e1aa3066811R518) and [here](https://github.com/ethereum/eth-account/pull/52/commits/aa9182811a67a673dc428f1f37ea41fad14754fc#diff-1de692d823115945c2232e1aa3066811R564).

carver commented on tests/core/test_accounts.py:
```diff
@@ -481,23 +507,30 @@ def test_eth_account_encrypt(
+        encrypted = account.encrypt(password, kdf=kdf, iterations=iterations)
 
     assert encrypted['address'] == '2c7536e3605d9c16a7a3d7b1898e529396a65c23'
     assert encrypted['version'] == 3
```
Just after the kdf assertion looks like the right place to verify the iterations in the encrypted keyfile json.

```
In [3]: create_keyfile_json(b'1'*32, b'aonteuhsantoheu', kdf='scrypt', iterations=128)
Out[3]: 
{'address': '77952ce83ca3cad9f7adcfabeda85bd2f1f52008',
 'crypto': {'cipher': 'aes-128-ctr',
  'cipherparams': {'iv': 'b38a0dc532c258ae55a08b1967a03e23'},
  'ciphertext': 'cbebc97f6ef6642fbc9e77bce055d41980ddae28ddf3e1cca95402035ccce94b',
  'kdf': 'scrypt',
  'kdfparams': {'dklen': 32,
   'n': 128,  # <---------------------------------------------------- iterations shows up here
   'p': 8,
   'r': 1,
   'salt': '27f214a3a0ca8b07f3168902e449417d'},
  'mac': 'bf22caaa85025c22b9081326f4e7646cb78ec369540ce31fe2c3130d67274a75'},
 'id': 'c7d68f4a-0b4b-460f-b128-6ef5b03816c1',
 'version': 3}

In [4]: create_keyfile_json(b'1'*32, b'aonteuhsantoheu', iterations=128)
Out[4]: 
{'address': '77952ce83ca3cad9f7adcfabeda85bd2f1f52008',
 'crypto': {'cipher': 'aes-128-ctr',
  'cipherparams': {'iv': '8fe849a2d481c224493fdb4e0284cfd5'},
  'ciphertext': 'b1f45f5ff9a81479237a241ccdc9efde073cb663d3461a06f2cd7ea0dfc487a0',
  'kdf': 'pbkdf2',
  'kdfparams': {'c': 128,  # <---------------------------------------- iterations shows up here
   'dklen': 32,
   'prf': 'hmac-sha256',
   'salt': '70dd2aa94c18583833b558814f1ce3fa'},
  'mac': '1d5b5d1977da7702e4592a0d8ce3c47da1fcb0ab15b055ac37e346c0a3bc011c'},
 'id': '0b9afb25-e29d-4ed6-ba51-8dfbb98c3362',
 'version': 3}

stefanmendoza replied:
Thanks for pointing that out! See comment here: https://github.com/ethereum/eth-account/pull/52/commits/aa9182811a67a673dc428f1f37ea41fad14754fc#r247003981

stefanmendoza commented on eth_account/signers/local.py:
```diff
@@ -48,15 +48,20 @@ def privateKey(self):
         :meth:`~eth_account.account.Account.encrypt`, but without a private key argument.
         '''
         if kdf is None:
             return self._publicapi.encrypt(self.privateKey, password)
```
```
tests/core/test_accounts.py::test_eth_account_prepared_encrypt[hex_str_default_kdf_provided_iterations] FAILED                                                                [ 68%]

...

        if expected_kdf == 'pbkdf2':
            assert encrypted['crypto']['kdfparams']['c'] == expected_iterations
        elif expected_kdf == 'scrypt':
>           assert encrypted['crypto']['kdfparams']['n'] == expected_iterations
E           assert 262144 == 1024
```
Good catch!

stefanmendoza replied:
Fixed: https://github.com/ethereum/eth-account/pull/52/commits/aa9182811a67a673dc428f1f37ea41fad14754fc#diff-0f68fdf3bef16bc9585bd0542646fe4aR56

stefanmendoza commented:
@carver I think I've addressed all your comments. Let me know whatcha think! Thanks for catching those, that was a miss. Also, glad to have you back from vacation 🎉

carver commented on tests/core/test_accounts.py:
```diff
@@ -419,90 +422,155 @@ def test_eth_account_recover_transaction_from_eth_test(acct, transaction):
+    if expected_kdf == 'pbkdf2':
+        assert encrypted['crypto']['kdfparams']['c'] == expected_iterations
+    elif expected_kdf == 'scrypt':
+        assert encrypted['crypto']['kdfparams']['n'] == expected_iterations
```
Cool, looks good. I just want to wrap this up with an else block that fails so that we don't accidentally skip testing iterations if we add a new kdf.

I'll add the commit, and we'll be good to go.

carver commented:
Thanks @stefanmendoza for your work! (and patience while I was out :))
